File: JbmrksIOs/JbmrksIOs/Models/Notification.swift
//
//  Notification.swift
//  JbmrksIOs
//
//  Notification model
//

import Foundation

enum NotificationType: String, Codable {
    case taskAssigned = "TASK_ASSIGNED"
    case taskUpdated = "TASK_UPDATED"
    case taskComment = "TASK_COMMENT"
    case taskDeadline = "TASK_DEADLINE"
    case taskStatusChanged = "TASK_STATUS_CHANGED"
    case fileAttached = "FILE_ATTACHED"
    case feedPost = "FEED_POST"
    case chatMessage = "CHAT_MESSAGE"
    case general = "GENERAL"
}

enum NotificationPriority: String, Codable {
    case low = "LOW"
    case normal = "NORMAL"
    case high = "HIGH"
    case urgent = "URGENT"
}

struct Notification: Identifiable, Codable {
    let id: String
    let type: NotificationType
    let title: String
    let message: String
    let timestamp: Int64
    let isRead: Bool
    let priority: NotificationPriority
    let relatedId: String?
    let actionUrl: String?
    
    func getFormattedTime() -> String {
        let now = Int64(Date().timeIntervalSince1970 * 1000)
        let diff = now - timestamp
        
        if diff < 60000 {
            return "Just now"
        } else if diff < 3600000 {
            return "\(diff / 60000) minutes ago"
        } else if diff < 86400000 {
            return "\(diff / 3600000) hours ago"
        } else if diff < 604800000 {
            return "\(diff / 86400000) days ago"
        } else {
            let date = Date(timeIntervalSince1970: Double(timestamp) / 1000)
            let formatter = DateFormatter()
            formatter.dateFormat = "MMM dd, yyyy"
            return formatter.string(from: date)
        }
    }
    
    func getIcon() -> String {
        switch type {
        case .taskAssigned: return "📋"
        case .taskUpdated: return "✏️"
        case .taskComment: return "💬"
        case .taskDeadline: return "⏰"
        case .taskStatusChanged: return "🔄"
        case .fileAttached: return "📎"
        case .feedPost: return "📰"
        case .chatMessage: return "💬"
        case .general: return "🔔"
        }
    }
}

File: app/src/main/java/com/example/jbmarks/notifications/domain/Notification.kt
package com.example.jbmarks.notifications.domain

import java.util.*

/**
 * Notification types
 */
enum class NotificationType {
    TASK_ASSIGNED,
    TASK_UPDATED,
    TASK_COMMENT,
    TASK_DEADLINE,
    TASK_STATUS_CHANGED,
    FILE_ATTACHED,
    FEED_POST,
    CHAT_MESSAGE,
    GENERAL
}

/**
 * Notification priority
 */
enum class NotificationPriority {
    LOW,
    NORMAL,
    HIGH,
    URGENT
}

/**
 * Domain model for a notification
 */
data class Notification(
    val id: String,
    val type: NotificationType,
    val title: String,
    val message: String,
    val timestamp: Long,
    val isRead: Boolean,
    val priority: NotificationPriority,
    val relatedId: String?, // ID of related task, comment, etc.
    val actionUrl: String? // Deep link or action to take
) {
    fun getFormattedTime(): String {
        val now = System.currentTimeMillis()
        val diff = now - timestamp
        
        return when {
            diff < 60000 -> "Just now"
            diff < 3600000 -> "${diff / 60000} minutes ago"
            diff < 86400000 -> "${diff / 3600000} hours ago"
            diff < 604800000 -> "${diff / 86400000} days ago"
            else -> {
                val date = Date(timestamp)
                java.text.SimpleDateFormat("MMM dd, yyyy", Locale.getDefault()).format(date)
            }
        }
    }
    
    fun getIcon(): String {
        return when (type) {
            NotificationType.TASK_ASSIGNED -> "📋"
            NotificationType.TASK_UPDATED -> "✏️"
            NotificationType.TASK_COMMENT -> "💬"
            NotificationType.TASK_DEADLINE -> "⏰"
            NotificationType.TASK_STATUS_CHANGED -> "🔄"
            NotificationType.FILE_ATTACHED -> "📎"
            NotificationType.FEED_POST -> "📰"
            NotificationType.CHAT_MESSAGE -> "💬"
            NotificationType.GENERAL -> "🔔"
        }
    }
}
